### PR TITLE
Add Tree Inference Code

### PR DESCRIPTION
**Pull request by jamesmcclain**
## Overview

An ML model has been added to perform tree health inference, along with associated code.

## Demo

image
```
argo submit -n argo workflow-treemodel-id.yaml -p 'activator-input=["f140603t01p00r08_sc01"]' --watch
```

## Notes

- The docker image is currently on `quay.io` instead of ECR
- ~~The ML process seems to produce very noticeably different results when run on a CPU as compared to a GPU.  The reason for this is under investigation.~~

## Checklist

- [x] CI passes after rebase
- [x] README.md updated if necessary to reflect the changes

### CODE REVIEW

File: src/pipeline/activator/activator/treehealth/tree.py [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+import torch.nn.functional as F
+
+
+class TreeModel(torch.nn.Module):
+    def __init__(self, preshrink: int):
+        super().__init__()
+        magic_number = 13
+        self.cheaplabs = torch.nn.ModuleList([
+            torch.hub.load(
+                'jamesmcclain/CheapLab:38af8e6cd084fc61792f29189158919c69d58c6a',
```
Comment (pomadchin) [inline src/pipeline/activator/activator/treehealth/tree.py:17]: would it be possible to put it under the azavea quay account or into the ECR?

File: src/pipeline/activator/requirements-conda.txt [hunk 2]
```diff
@@ -8,4 +8,6 @@ tqdm==4.61.2
 h5py==3.3.0
 pyproj==3.1.0
 shapely==1.7.1
-rasterio==1.2.1
\ No newline at end of file
+rasterio==1.2.1
+pytorch-cpu=1.9.0
+torchvision-cpu=0.9.1
```
Comment (pomadchin) [inline src/pipeline/activator/requirements-conda.txt:13]: I'm wondering, how heavy this deps are? If they are not let's keep them as-is, if they are: would it make sense to have a separate docker image?

**Review by pomadchin — COMMENTED** [review]
Let's mb merge it as is once there are no conflicts; looks cool 💯 

I see that this PR changes the docker image deps, which will be published by the CI to ECR (or it is possible to do it manually via `./docker/ecr-publish.sh`) Is it possible to use ECR docker images everywhere rather than personal quay.io?

File: src/pipeline/activator/activator/treehealth/main.py [hunk 3]
```diff
@@ -0,0 +1,251 @@
...
+            k2 = to_snake(k)
+            setattr(args, k2, args_json.get(k))
+    return args
+
+
+def cli_parser():
+    parser = argparse.ArgumentParser()
+
+    parser.add_argument("--aviris-collection-id", type=str, default=AVIRIS_ARCHIVE_COLLECTION_ID)
+    parser.add_argument("--aviris-stac-id", type=str, default="f140603t01p00r18_sc01", help="STAC Item ID to process from the STAC collection")
```
Comment (pomadchin) [inline src/pipeline/activator/activator/treehealth/main.py:124]: Should not be this abstracted over the sensor type, or you're pretty sure that it is very AVRIS specific?

Comment (jamesmcclain) [reply src/pipeline/activator/activator/treehealth/main.py:124]: This is decidedly AVIRIS specific.  Can revisit later if we add models for other imagery types.

**Review by pomadchin — APPROVED** [review]
Lets merge it as is, it is good enough :+1: and looks really good; my only concern is in the image names, I'd like to see them referring to the ECR.

**Comment by jamesmcclain** [comment]
@jpolchlo Still need a review for this PR

File: deployment/argo-workflows/eks/workflow-treemodel-scene-id.yaml [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+    retryStrategy:
+      limit: "10"
+      expression: "asInt(lastRetry.exitCode) == -42"
+      # https://argoproj.github.io/argo-workflows/fields/#backoff
+      backoff:
+        duration: "10m" # Default unit is seconds. Could also be a Duration, e.g.: "2m", "6h", "1d"
+        factor: "2"
+        maxDuration: "1d"
+    container:
+      image: quay.io/jmcclain/activator:latest
```
Comment (jpolchlo) [inline deployment/argo-workflows/eks/workflow-treemodel-scene-id.yaml:39]: What's the nature of this container?  Is it general enough to land in an Azavea-owned repository?

Comment (jamesmcclain) [reply deployment/argo-workflows/eks/workflow-treemodel-scene-id.yaml:39]: Yes, as discussed above, the plan is to push it to ECR after PR approval.

**Review by jpolchlo — APPROVED** [review]
Minor comments, otherwise looks good.

**Comment by jpolchlo** [comment]
Is the training process and data selection for this model fully in #210 ?

File: src/pipeline/activator/activator/treehealth/main.py [hunk 5]
```diff
@@ -0,0 +1,251 @@
...
+        warpMemoryLimit = int(os.environ.get("GDAL_WARP_MEMORY_LIMIT", None))
+    except TypeError:
+        warpMemoryLimit = None
+
+    args = cli_parser().parse_args()
+
+    if not args.aviris_stac_id.startswith(args.aviris_collection_id):
+        args.aviris_stac_id = args.aviris_collection_id + '_' + args.aviris_stac_id
+
+    s3 = boto3.client("s3")
```
[the remark this answers is not in the document]
Comment (jamesmcclain) [reply src/pipeline/activator/activator/treehealth/main.py:150]: :heavy_check_mark:

File: deployment/argo-workflows/eks/workflow-treemodel-scene-id.yaml (hunk 4 above)
Comment (jamesmcclain) [reply deployment/argo-workflows/eks/workflow-treemodel-scene-id.yaml:39]: :heavy_check_mark:

File: research/src/algae-model/src/cheaplab_view.py [hunk 6]
```diff
@@ -60,13 +58,16 @@ def cli_parser():
         from cloud import make_cloud_model
         model = make_cloud_model(in_channels=[224], preshrink=1)
     elif args.architecture == 'tree':
-        from tree import make_tree_model
+        try:
+            from .tree import make_tree_model
+        except:
+            from tree import make_tree_model
```
[the remark this answers is not in the document]
Comment (jamesmcclain) [reply research/src/algae-model/src/cheaplab_view.py:64]: :heavy_check_mark:

**Comment by jamesmcclain** [comment]
> Is the training process and data selection for this model fully in #210 ?

Good call, I should add documentation for this in a future PR.

File: src/pipeline/activator/activator/treehealth/main.py [hunk 7]
```diff
@@ -0,0 +1,251 @@
...
+        os.system(f'aws s3 cp {args.pipeline_uri} {filename_pipeline}')
+    elif args.pipeline:
+        filename_pipeline = args.pipeline
+    if filename_pipeline is not None:
+        with open(filename_pipeline, 'r') as file:
+            args_json = json.loads(file.read().replace('\n', ''))
+        for k in args_json:
+            k2 = to_snake(k)
+            setattr(args, k2, args_json.get(k))
+    return args
```
[the remark this answers is not in the document]
Comment (jamesmcclain) [reply src/pipeline/activator/activator/treehealth/main.py:117]: Good catch, that function is needed to integrate arguments made by the pipeline.  I accidentally omitted the line that uses it.

**Pull request merged**